File: Doc/Kattan/Solutions-Manual/problem_2_1.m
% =========================================================================
% Problem 2.1 - Two-Element Spring System (Fig. 2.4)
% Reference: P. I. Kattan, "MATLAB Guide to Finite Elements: An Interactive
% Approach" - Chapter 2, spring element formulation.
%
% System:  1 --[k1]-- 2 --[k2]-- 3
%          Node 1 and Node 3 are fixed to walls (u1 = u3 = 0)
%          Force P is applied at Node 2, along the 1->3 axis
%
% Computes:
%   1. Global stiffness matrix K
%   2. Displacement at node 2
%   3. Reactions at nodes 1 and 3
%   4. Force in each spring
%
% Requires springElementStiffness.m, springAssemble.m, springElementForce.m
% (in the same folder as this script).
% =========================================================================

clear; clc;

% ---------------------- Given data --------------------------------------
k1 = 200;   % spring 1 stiffness, kN/m
k2 = 250;   % spring 2 stiffness, kN/m
P  = 10;    % applied force at node 2, kN

n_nodes = 3;

% ---------------------- 1. Assemble global stiffness matrix -------------
K = zeros(n_nodes);

k1_local = springElementStiffness(k1);
K = springAssemble(K, k1_local, 1, 2);

k2_local = springElementStiffness(k2);
K = springAssemble(K, k2_local, 2, 3);

disp('=========================================================');
disp('1. Global stiffness matrix K (kN/m):');
disp(K);

% ---------------------- 2. Apply boundary conditions and solve ----------
% u1 = 0 (fixed), u3 = 0 (fixed) -> node 2 is the only free DOF
free_dof = 2;

F = zeros(n_nodes,1);
F(2) = P;

u = zeros(n_nodes,1);
u(free_dof) = K(free_dof,free_dof) \ F(free_dof);

disp('=========================================================');
fprintf('2. Nodal displacements:\n');
fprintf('   u1 = %.6f m (fixed)\n', u(1));
fprintf('   u2 = %.6f m\n', u(2));
fprintf('   u3 = %.6f m (fixed)\n', u(3));

% ---------------------- 3. Reactions at nodes 1 and 3 -------------------
% Full-system equilibrium: K*u = F + R   =>   R = K*u - F
R = K*u - F;

disp('=========================================================');
fprintf('3. Reactions:\n');
fprintf('   R1 = %.4f kN\n', R(1));
fprintf('   R3 = %.4f kN\n', R(3));

% ---------------------- 4. Force in each spring --------------------------
f1 = springElementForce(k1, u(1), u(2));
f2 = springElementForce(k2, u(2), u(3));

disp('=========================================================');
fprintf('4. Element (spring) forces (+ = tension, - = compression):\n');
fprintf('   f1 (spring 1, nodes 1-2) = %.4f kN\n', f1);
fprintf('   f2 (spring 2, nodes 2-3) = %.4f kN\n', f2);

disp('=========================================================');
fprintf('Equilibrium check, R1 + P + R3 = %.6f (should be ~0)\n', R(1)+P+R(3));

File: problems/problem_2_1.jl
#!/usr/bin/env julia
# ═══════════════════════════════════════════════════════════════
# Problem 2.1 — Two-Element Spring System (Fig. 2.4)
# Reference: P. I. Kattan, "MATLAB Guide to Finite Elements:
#   An Interactive Approach" (2nd ed., Springer, 2007)
# ═══════════════════════════════════════════════════════════════
#
#                             P
#               k1          ----->         k2
# |/----o-----/\/\/\----------o---------/\/\/\-----o----\|
# |/    1                     2                    3    \|
# |/                                                    \|
#
# ═══════════════════════════════════════════════════════════════
# Computes:
#   1. Global stiffness matrix K
#   2. Displacement at node 2
#   3. Forces (reactions at 1,3 and internal spring forces)
#   4. Equilibrium check
# ═══════════════════════════════════════════════════════════════

using LibFEM
using LinearAlgebra

# ─── Parameters ──────────────────────────────────────────────
k1_val = 200.0
k2_val = 250.0

# ─── Element stiffness matrices ──────────────────────────────
k1 = d1_spring_elementstiffness(k1_val)
k2 = d1_spring_elementstiffness(k2_val)

println("k1 =")
display(k1)
println("k2 =")
display(k2)

# ─── Assembly ────────────────────────────────────────────────
K = zeros(3, 3)
K = d1_spring_assemble(K, k1, 1, 2)
K = d1_spring_assemble(K, k2, 2, 3)

println("\nK =")
display(K)

# ─── Solve ───────────────────────────────────────────────────
k = K[2:2, 2:2]
f = [10.0]

u = k \ f
U = [0.0; u; 0.0]
F = K * U

println("\nk =")
display(k)
println("\nf =")
display(f)
println("\nu =")
display(u)
println("\nU =")
display(U)
println("\nF =")
display(F)

# ─── Post-processing: element forces ─────────────────────────
u1 = [0.0; u]
f1 = d1_spring_elementforce(k1, u1)

u2 = [u; 0.0]
f2 = d1_spring_elementforce(k2, u2)

println("\nu1 =")
display(u1)
println("\nf1 =")
display(f1)
println("\nu2 =")
display(u2)
println("\nf2 =")
display(f2)

# ─── Equilibrium check ───────────────────────────────────────
println("\n--- Equilibrium check ---")
println("Sum of reactions: ", sum(F[1:3:end]))
println("Applied force P: 10.0")
println("Reacted force at node 1 (F1): ", F[1])
println("Reacted force at node 3 (F3): ", F[3])
println("Spring force f1: ", f1)
println("Spring force f2: ", f2)
println("Sum F1 + F3 = ", F[1] + F[3], " (should equal -10.0)")
println("Equilibrium satisfied (round-trip): ", abs(F[1] + F[3] + 10.0) < 1e-10)

# ─── Self-validation ─────────────────────────────────────────
# Expected values verified against Octave execution of Kattan's problem_2_1.m
@assert isapprox(K, [200 -200 0; -200 450 -250; 0 -250 250]; rtol=1e-10) "K mismatch"
@assert isapprox(k, [450.0]; rtol=1e-10) "k mismatch"
@assert isapprox(f, [10.0]; rtol=1e-10) "f mismatch"
@assert isapprox(u, [0.022222222222222223]; rtol=1e-10) "u mismatch"
@assert isapprox(U, [0.0; 0.022222222222222223; 0.0]; rtol=1e-10) "U mismatch"
@assert isapprox(F, [-4.444444444444445; 10.0; -5.555555555555555]; rtol=1e-10) "F mismatch"
@assert isapprox(f1, [-4.444444444444445; 4.444444444444445]; rtol=1e-10) "f1 mismatch"
@assert isapprox(f2, [5.555555555555555; -5.555555555555555]; rtol=1e-10) "f2 mismatch"
